subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/WatchingNotSupportedVirtualFileSystem.java

```diff
@@ -46,14 +49,19 @@ public WatchingNotSupportedVirtualFileSystem(VfsRootReference rootReference) {
         this.rootReference = rootReference;
     }
 
+    @Override
+    public boolean isWatchingFileSystem() {
+        return false;
+    }
+
     @Override
     protected SnapshotHierarchy updateNotifyingListeners(UpdateFunction updateFunction) {
         return updateFunction.update(SnapshotHierarchy.NodeDiffListener.NOOP);
     }
 
     @Override
-    public void afterBuildStarted(boolean watchingEnabled, VfsLogging vfsLogging, WatchLogging watchLogging, BuildOperationRunner buildOperationRunner) {
-        if (watchingEnabled) {
+    public void afterBuildStarted(WatchMode watchMode, VfsLogging vfsLogging, WatchLogging watchLogging, BuildOperationRunner buildOperationRunner) {
+        if (watchMode == WatchMode.ENABLED) {
             LOGGER.warn("Watching the file system is not supported on this operating system.");
```

Comment:
I'm just wondering if we shouldn't just fail in this case. You ask for file watching but it's not available. The warning can be lost in the logs...

Reply:
This is a bit more complicated than I first thought. There are a number of reasons not to be able to enable watching:

1) it's not supported on the current OS
2) there's a problem with starting it at the start of the build, like `inotify` count running out
3) we might be able to start it, but we have to stop it during the build at some point, because `inotify` runs out then, or there's an error reported by the OS

Keep in mind that most people will just use the default, and will either not even know about the feature, or expect it to work fine in all their builds.

Let's keep things as they are for now, and come back when we have a way to better expose warnings during a build.

Reply:
I would prefer to fail the build when file-system watching is explicitly enabled and we don't enable because of 1). 2) and 3) happen during the build, so they have nothing to do with the flag being enabled or not - it is a problem with the optimization. IMO, 2) and 3) should be warnings and should be handled how we handle other warnings. A similar thing happens when a download from the remote build cache failed, I think, and that also should have no effect on the state of the flag shown in build scans or failing the build.

Having said all that I leave it up to you whether to fail the build in case 1) or not as part of this PR.

Reply:
I'd go with failing here but for one reason. If they use explicitly enables the feature, and it is supported on the Linux they are running, but we can't enable it anyway because of `inotify` handle counts, from the user's point of view that's quite a similar situation to the OS not being supported. In both cases they likely need to go through system administrators to either upgrade the OS, or to bump the number of `inotify` handles. Yet we'd be failing the build in one case, and showing a warning in another.

I think it's fine to merge this PR without resolving this issue and we should address this question separately.